src/nvim/os_unix.c

```diff
@@ -186,6 +186,7 @@ void mch_exit(int r)
   free_all_mem();
 #endif
 
+  stream_set_blocking(fileno(stdin), true);
```

Comment:
`fileno(stdin)` is not always the file descpritor to unblock. With the following hack `echo foo|nvim ... && read` works:
```diff
diff --git a/src/nvim/main.c b/src/nvim/main.c
index 5f0372f..452c23f 100644
--- a/src/nvim/main.c
+++ b/src/nvim/main.c
@@ -180,6 +180,7 @@ void early_init(void)
   set_lang_var();               // set v:lang and v:ctype
 }
 
+int global_fd;
 #ifdef MAKE_LIB
 int nvim_main(int argc, char **argv)
 #else
@@ -282,6 +283,8 @@ int main(int argc, char **argv)
       fd = params.err_isatty ? fileno(stderr) : fileno(stdout);
     }
     input_start_stdin(fd);
+    fprintf(stderr,"input_start_stdin(%d)\n",fd);
+    global_fd=fd;
   }
 
   // open terminals when opening files that start with term://
diff --git a/src/nvim/os/stream.c b/src/nvim/os/stream.c
index 7d11997..30026cd 100644
--- a/src/nvim/os/stream.c
+++ b/src/nvim/os/stream.c
@@ -18,6 +18,6 @@ int stream_set_blocking(int fd, bool blocking)
   uv_pipe_init(uv_default_loop(), &stream, 0);
   uv_pipe_open(&stream, fd);
   int retval = uv_stream_set_blocking((uv_stream_t *)&stream, blocking);
-  uv_close(&stream, NULL);
+  uv_close((uv_handle_t *)&stream, NULL);
   return retval;
 }
diff --git a/src/nvim/os_unix.c b/src/nvim/os_unix.c
index 9d4e7b8..5cf64e3 100644
--- a/src/nvim/os_unix.c
+++ b/src/nvim/os_unix.c
@@ -186,7 +186,8 @@ void mch_exit(int r)
   free_all_mem();
 #endif
 
-  stream_set_blocking(fileno(stdin), true);
+  extern int global_fd;
+  stream_set_blocking(global_fd, true);
   exit(r);
 }
 

```